File: Leetcode/numTrees.swift
var cache: [Int: Int] = [:]
    func numTrees(n: Int) -> Int {
        if let num = cache[n] {
            return num
        }
        if n == 0 || n == 1 {
            cache[0] = 1
            cache[1] = 1
            return 1
        } else {
            var total: Int = 0
            for var i in 1 ... n {
                //left 1 ... i -1  i-1
                //right i+1 ... n  n-i 
                total += numTrees(i-1) * numTrees(n-i)
            }
            cache[n] = total
            return total
        }
    }

File: Leetcode/numTrees.java
public int numTrees(int n) {
        if (cache.containsKey(n)) {
            return cache.get(n);
        }
        if (n == 0 || n == 1) {
            cache.put(n, 1);
            return 1;
        }
	int total = 0;
        for (int i = 1; i <= n; i++) {
            //1.. i-1
            //i+1, n
            int left = numTrees(i - 1);
            int right = numTrees(n - i);
            total += left * right;
        }
        cache.put(n, total);
        return total;
    }
